core/chains/evm/client/node_fsm.go

```diff
@@ -0,0 +1,218 @@
+	case NodeStateOutOfSync:
+		return "OutOfSync"
+	case NodeStateClosed:
+		return "Closed"
+	default:
+		return fmt.Sprintf("NodeState(%d)", n)
+	}
+}
+
+// GoString prints a prettier state
+func (n NodeState) GoString() string {
+	return fmt.Sprintf("NodeState%s(%d)", n.String(), n)
+}
+
+const (
+	// NodeStateUndialed is the first state of a virgin node
+	NodeStateUndialed = NodeState(iota)
+	// NodeStateDialed is after a node has successfully dialed but before it has verified the correct chain ID
+	NodeStateDialed
+	// NodeStateInvalidChainID is after chain ID verification failed
+	NodeStateInvalidChainID
+	// NodeStateAlive is a healthy node after chain ID verification succeeded
+	NodeStateAlive
+	// NodeStateUnreachable is a node that cannot be dialed or has disconnected
+	NodeStateUnreachable
+	// NodeStateOutOfSync is a node that is accepting connections but exceeded
+	// the failure threshold without sending any new heads. It will be
+	// disconnected, then put into a revive loop and re-awakened after redial
+	// if a new head arrives
+	NodeStateOutOfSync
+	// NodeStateClosed is after the connection has been closed and the node is at the end of its lifecycle
+	NodeStateClosed
+	// nodeStateLen tracks the number of states
+	nodeStateLen
+)
+
+// allNodeStates represents all possible states a node can be in
+var allNodeStates []NodeState
+
+func init() {
+	for s := NodeState(0); s < nodeStateLen; s++ {
+		allNodeStates = append(allNodeStates, s)
+	}
+
+}
+
+// FSM methods
+
+// State allows reading the current state of the node
+func (n *node) State() NodeState {
+	n.stateMu.RLock()
+	defer n.stateMu.RUnlock()
+	return n.state
+}
+
+// setState is only used by internal state management methods.
+// This is low-level; care should be taken by the caller to ensure the new state is a valid transition.
+// State changes should always be synchronous: only one goroutine at a time should change state.
+// n.stateMu should not be locked for long periods of time because external clients expect a timely response from n.State()
+func (n *node) setState(s NodeState) {
+	n.stateMu.Lock()
+	defer n.stateMu.Unlock()
+	n.state = s
+}
+
+// declareXXX methods change the state and pass conrol off the new state
+// management goroutine
+// TODO: unit test the state transitions
+
+func (n *node) declareAlive() {
+	n.stateMu.Lock()
+	defer n.stateMu.Unlock()
+	if n.state == NodeStateClosed {
+		return
+	}
+	n.transitionToAlive()
+	n.log.Infow("RPC Node is online", "nodeState", n.State())
+	n.wg.Add(1)
+	go n.aliveLoop()
+}
+
+func (n *node) transitionToAlive() {
+	switch n.state {
+	case NodeStateDialed:
+		n.state = NodeStateAlive
+	default:
+		panic(fmt.Sprintf("cannot transition from %#v to %#v", n.state, NodeStateAlive))
+	}
+}
+
+// declareInSync puts a node back into Alive state, allowing it to be used by
+// pool consumers again
+func (n *node) declareInSync() {
+	n.stateMu.Lock()
+	defer n.stateMu.Unlock()
+	if n.state == NodeStateClosed {
+		return
+	}
```

Comment:
Would it be helpful for testing and readability to encapsulate the pure fsm with a simple type that is embedded in node?
```
type nodeFSM struct{
    state, stateMu ...
}
```
Tradeoff might be passing anonymous closures for things that need to be done while the lock is held.

Reply:
Actually, maybe just method references instead of anonymous closures? Might end up pretty clean actually :shrug:

Reply:
Could you provide an example maybe? Not sure I follow your suggestion

Reply:
Something like this:
```go
func (n *node) declareInSync() {
	n.transitionToInSync(func() {
		n.log.Infow("RPC Node is back in sync", "nodeState", n.state)
		n.wg.Add(1)
		go n.aliveLoop()
	})
}

func (n *nodeFSM) transitionToInSync(fn func()) {
	n.stateMu.Lock()
	defer n.stateMu.Unlock()
	if n.state == NodeStateClosed {
		return
	}
	switch n.state {
	case NodeStateOutOfSync:
		n.state = NodeStateAlive
	default:
		panic(fmt.Sprintf("cannot transition from %#v to %#v", n.state, NodeStateAlive))
	}
	fn()
}
```
And then in some cases you could do `n.transitionToInSync(n.methodName)` instead of closures.
:shrug: